pallets/attestation/src/migrations.rs

```diff
@@ -0,0 +1,133 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2023 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+use frame_support::{
+	traits::{Get, GetStorageVersion, OnRuntimeUpgrade, ReservableCurrency, StorageVersion},
+	weights::Weight,
+};
+use kilt_support::migration::switch_reserved_to_hold;
+use log;
+use sp_runtime::SaturatedConversion;
+use sp_std::marker::PhantomData;
+
+use crate::{AccountIdOf, Attestations, Config, CurrencyOf, HoldReason, Pallet};
+
+pub struct BalanceMigration<T>(PhantomData<T>);
+
+impl<T: crate::pallet::Config> OnRuntimeUpgrade for BalanceMigration<T>
+where
+	<T as Config>::Currency: ReservableCurrency<T::AccountId>,
+{
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		log::info!("Attestation: Initiating migration");
+
+		let onchain_storage_version = Pallet::<T>::on_chain_storage_version();
+		if onchain_storage_version.eq(&StorageVersion::new(1)) {
+			StorageVersion::new(2).put::<Pallet<T>>();
+			return do_migration::<T>();
+		}
+
+		log::info!(
+			"Attestation: No migration needed. This file should be deleted. Current storage version: {:?}, Required Version for update: {:?}", 
+			onchain_storage_version,
+			StorageVersion::new(1)
+		);
+
+		<T as frame_system::Config>::DbWeight::get().reads_writes(0, 0)
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<sp_std::vec::Vec<u8>, &'static str> {
+		use sp_std::vec;
+
+		let has_all_user_no_holds = Attestations::<T>::iter_values()
+			.map(|details: crate::AttestationDetails<T>| {
+				kilt_support::migration::has_user_reserved_balance::<AccountIdOf<T>, CurrencyOf<T>>(
+					&details.deposit.owner,
+					&T::RuntimeHoldReason::from(HoldReason::Deposit),
+				)
+			})
+			.all(|user| user);
+
+		assert!(
+			has_all_user_no_holds,
+			"Pre Upgrade Attestation: there are users with holds!"
+		);
+
+		assert_eq!(Pallet::<T>::on_chain_storage_version(), StorageVersion::new(1));
```

Comment:
Same here with the storage version. Since `new` is a `const`, maybe it can be pulled out and used everywhere in this migration.